readevent

kashish2210 opened the pull request:
#27 
my [workfloww](https://github.com/kashish2210/notebooks/blob/main/eventlist.ipynb)
i have tested using [kplr011446443.fits](https://archive.stsci.edu/pub/kepler/lightcurves/)
for direct getting this FITS file you can visit my [github](https://github.com/kashish2210/notebooks/blob/main/kplr011446443.fits)

further I have created 21 test cases that includes :
- to test monol_testA.evt from `test/data`

@fjebaker @JawadAhmadCS your review required :)

kashish2210 commented:
I need a suggestion of fits that I should use that has `energy values`

kashish2210 commented:
I was thinking of adding the case for .txt and .gz files but as per your comment I have done minimal implementation in future we can add these if needed

fjebaker commented on src/events.jl:
```diff
@@ -0,0 +1,43 @@
+            end
+        end
+    end
+
```
I'd suggest adding some kind of warning or even error if the `times` and `energies` vectors are still empty at this point, so the caller knows that their FITS file may have been invalid.

fjebaker commented on src/Stingray.jl:
```diff
@@ -32,4 +32,7 @@ export bin_intervals_from_gtis
 
-end 
+include("events.jl")
+export readevents
```
```suggestion
export EventList, Meta, readevents
```

fjebaker commented on src/events.jl:
```diff
@@ -0,0 +1,43 @@
+struct Meta
```
I'd maybe suggest renaming this struct to something like `DictMetadata`?

fjebaker reviewed (commented):
Hi @kashish2210 this looks great! Just two little comments and a suggestion to export your structs.

fjebaker commented:
Also, please could you add docstrings and struct field documentation! That will help us maintain this code going forwards :+1:

kashish2210 commented:
yup I have added the necessary docstrings and warns and updated the test cases to suppress warns while testing for advance implementations we will include `logconfig.jl`

fjebaker commented on src/events.jl:
```diff
@@ -0,0 +1,90 @@
+    energies::Vector{T}
+    metadata::DictMetadata
+end
+
+function readevents(path; T = Float64)
+    """
+    Read event data from a FITS file into an EventList structure.
+    Parameters
+    ----------
+    path : String
+        Path to the FITS file.
+    Keyword Arguments
+    ----------------
+    T : DataType
+        The numeric type for time and energy values. Default is Float64.
+    Returns
+    -------
+    EventList{T}
+        An EventList structure containing the extracted data.
+    Notes
+    -----
+    The function extracts TIME and ENERGY columns from any TableHDU in the FITS file.
+    All headers from each HDU are collected into the metadata field.
+    """
```
Julia function docstrings go above the function definition. It is also more
standard to document the parameters in prose instead of as lists, as much of
their semantics is in the function signature.

```suggestion

"""
    readevents(path; T = Float64)

Read event data from a FITS file into an EventList structure. The `path` is a
string that points to the location of the FITS file. `T` is used to specify
which numeric type to convert the data to.

Returns an [`EventList`](@ref) containing the extracted data.

## Notes

The function extracts `TIME` and `ENERGY` columns from any TableHDU in the FITS
file. All headers from each HDU are collected into the metadata field.
"""
function readevents(path; T = Float64)
```

fjebaker commented on src/events.jl:
```diff
@@ -0,0 +1,90 @@
+
+A structure containing event data from a FITS file.
+
+Fields
+------
```
```suggestion
## Fields

```

fjebaker reviewed (changes requested):
Thanks for adding the docstrings! I have added some suggestion for how these are usually done in Julia.

kashish2210 commented:
> Thanks for adding the docstrings! I have added some suggestion for how these are usually done in Julia.

thanks for suggestions

fjebaker commented on src/events.jl:
```diff
@@ -0,0 +1,85 @@
+
+A structure containing metadata from FITS file headers.
+
+Fields
+------
```
Sorry I missed just this one:

```suggestion
## Fields

```

fjebaker reviewed (approved):
Brilliant! Nice work @kashish2210

kashish2210 commented:
> Brilliant! Nice work @kashish2210

thanks :)
